roxie/udplib/udptopo.cpp

```diff
@@ -393,17 +436,28 @@ void TopologyManager::setRoles(const std::vector<RoxieEndpointInfo> &myRoles)
         topoBuf.append(role.channel).append('|');
         role.ep.getUrlStr(topoBuf);
         topoBuf.append('|').append(role.replicationLevel);
+        topoBuf.append('\t').append((__uint64) myInstance);
         topoBuf.append('\n');
     }
-    Owned<const ITopologyServer> newServer = new CTopologyServer(topoBuf);   // We set the initial topology to just the local information we know about
+}
+
+void TopologyManager::setRoles(const std::vector<RoxieEndpointInfo> &myRoles)
+{
+    topoBuf.clear();
+    myInstance = time(nullptr);
+    if (!myInstance) myInstance++;
+    _setRoles(myRoles, false);
+    Owned<const ITopologyServer> newServer = new CTopologyServer(topoBuf, nullptr);   // We set the initial topology to just the local information we know about
     SpinBlock b(lock);
     currentTopology.swap(newServer);
 }
 
-void TopologyManager::closedown()
+void TopologyManager::closedown(const std::vector<RoxieEndpointInfo> &myRoles)
 {
-    topoBuf.replaceString("server|", "-server|");
-    topoBuf.replaceString("agent|", "-agent|");
+    topoBuf.clear();
+    _setRoles(myRoles, true);
+    myInstance = 0;
```

Comment:
I'm not sure why there are two calls.  Could do with more explanation.

Reply:
The first one generates the text to ask the server to remove the old information. The second generates the text to give toposerver the new information (with instance=0).

Reply:
Could probably drop the second if the server treated a "remove" as setting the instance to 0